### ISSUE TITLE
Rename docker volume

### ISSUE THREAD

**Issue by leo90skk** [docker/docker#31154]
I miss a command which I can use to rename a volume.

Once I have created a lot of modules and forget to name the volumes at the start, it gets really difficult to identify which volume contains what.

A command like
`docker volume mv 27722c3065e2f074bc7f29963fe62f73669cb369617482596e222f9991daee57 my-nice-volume-name`
would be nice.
If that is not possible (because the id is used in a definition), it would be nice to have the ability to add a short comment:
`docker volume set-comment 27722c3065e2f074bc7f29963fe62f73669cb369617482596e222f9991daee57 "This is the volume /var/logs/ for my creat app"`
This comment should be visible in the list `docker volume ls`.

**Comment by franklinyu** [comment]
It's useful to rename a volume when using Docker-Compose, for example I can rename my `proj_data` to `proj_data_backup`, so that subsequent `docker-compose up` creates new `proj_data` volume, while I can still access the old `proj_data_backup`. Currently I have to edit my `docker-compose.yaml`.

**Comment by Happyholic1203** [comment]
need volume rename +1

**Comment by navanjr** [comment]
+1

**Comment by shishir-a412ed** [comment]
@thaJeztah I can open a PR for this. #dibs

Shishir

**Comment by thaJeztah** [comment]
@shishir-a412ed I'm not sure this one is easy to implement; it would require the path (on disk) of the container to be renamed, and given that multiple containers can use the same volume, those containers will have to be updated.

Possibly the solution would be to give volumes an _id_, and use that _id_ for the path (and store the volume _name_ as metadata of the volume)

@cpuguy83 any thoughts?

**Comment by franklinyu** [comment]
@thaJeztah Is it possible to check whether a volume is not linked to any (running or stopped) containers? It suffices to be able to rename stand-alone volumes.

**Comment by shishir-a412ed** [comment]
What I am implementing right now will allow you to rename a volume only if it is not used by running containers. I am guessing this API should give me that information:
https://github.com/moby/moby/blob/master/volume/store/errors.go#L52
I see your point, that stopped containers will have volume information/name in their metadata too which might needs to be updated.

Right now I was just implementing this change for local volumes. I was planning to extend that later to support volume plugins .e.g. If someone did `docker volume create foobar`.

docker would create a local directory under /var/lib/docker/volumes/foobar which will be bind mounted at `docker run` time. If the container is not running this is just a directory. 

`docker volume rename foobar newFoobar` would rename the directory to `/var/lib/docker/volumes/myFoobar` and update the `volumeStore` with this info.

This is what I am thinking. Let me know your thoughts:

1) `docker volume NAME NEW_NAME`
2) If the volume is in use by running containers, error out saying stop the containers first before renaming the volume.
3) If the volume is not used by running containers, we change the name of the volume directory (path on the host for either volume plugins or local volumes) and update the in memory stores which are doing the book keeping. We need to think about metadata on disk for stopped containers, that have old volume information.

This would require changes at:
1) docker/cli which would need daemon changes to be vendored in. So daemon PR should be merged first. (I already have this, I can open a PR if you like)
2) docker engine changes for handling `volume rename` 
3) Volume plugin API changes for handling `rename`

@thaJeztah @cpuguy83 thoughts ?

Shishir

**Comment by johanbrandhorst** [comment]
Any progress on this?

**Comment by cpuguy83** [comment]
This is hard. If we internally use an alias but still refer to the volume by the original name with the driver this could work, but I see this as a little strange.

**Comment by franklinyu** [comment]
@johanbrandhorst From #33966:

>  I think @thaJeztah wanted to support the use case for downgrading to an earlier version of docker. This design does not support that use case.
>
> I don't see an easy solution to this. so ideas are welcome.
> Also, feel free to take this over, if someone from the community wants to take this forward.

**Comment by bugdanov** [comment]
I ended up by creating a new volume and moving manually the content of /var/lib/docker/volumes/old-volume/_data/

**Comment by akibukis** [comment]
Same here. But rename option would be great, though.

**Comment by Hyruu** [comment]
For those who don't run Docker on Linux, and as such can't easily access /var/lib/docker on the root VM, you can still just create the new volume and create a temporary container to copy your data:

```
docker volume create --name <new_volume>
docker run --rm -it -v <old_volume>:/from -v <new_volume>:/to alpine ash -c "cd /from ; cp -av . /to"
docker volume rm <old_volume>
```

**Comment by franklinyu** [comment]
@Hyruu I’m tempted to write a bash function with completion…

**Comment by honoref** [comment]
That's work fine. 
I only add a test about the run command. I hope if the copy is wrong, errorlevel is greater than 0

#/bin/bash
docker volume create --name $2
docker run --rm -it -v $1:/from -v $2:/to alpine ash -c "cd /from ; cp -av . /to"
[ $? -eq 0 ] && docker volume rm $1

**Comment by moribellamy** [comment]
i used the docker clone script from https://github.com/gdiepen/docker-convenience-scripts.

1) docker_clone_volume.sh my-volume my-volume-bak
2) docker volume rm my-volume
3) docker_clone_volume.sh my-volume-bak my-volume

kinda roundabout but it works

**Comment by vedmant** [comment]
Sometimes docker makes me cry when it doesn't have such an obvious features. I don't think it's so complex to add, but without it project looks immature.

**Comment by thaJeztah** [comment]
@vedmant contributions welcome

**Comment by ruslanguns** [comment]
+1  I want this feature too

**Comment by gbrault** [comment]
+1 I want this feature too

**Comment by UncleSnycs** [comment]
+1 Me too

**Comment by zackishome** [comment]
+1

**Comment by jhwinter** [comment]
+1

**Comment by thaJeztah** [comment]
In case someone wants to work on this; I left some thoughts for a possible implementation in https://github.com/moby/moby/pull/33966#issuecomment-313821356 at the time, but (as may become apparent from that comment, and the discussion before/after it), implementing this is not _trivial_. It's essential that this feature does not cause a breaking change, and that upgrading (and preferably _downgrading_) works, and all those scenarios are part of the design, and solid.

Copying that comment here (but as mentioned; refer to the linked PR for more details);

My thought was to;

1. Implement a new storage scheme for volumes

May need to have @cpuguy83's input on this; I think there's an in-memory cache / index of volumes, and a boltdb store; I'm not sure if the `boltdb` store is "leading" or only used to attach meta-data.

This needs some thinking, given that we may have to support downgrading to an earlier version of Docker;

Change the volume store, so that each volume has both an `ID` and a `Name`;

- `ID` is a _unique_ index, under no circumstances two volumes can have the same `ID`. `ID` is a string, and for new volumes will contain a generated `ID`
- `Name` is also unique (question for @cpuguy83  - do we support two volumes with the same name, but different drivers?)

2. have a migration step

- get all the volumes that are known by the daemon
- for each volume (assuming volume-names are unique); copy the `Name` to `ID`. Given that `ID` is a string, the volume name would be a valid `ID`

Existing containers refer to volumes by their "Name", but now that we copied the `ID` to `Name`, we can have them refer to volumes by their `ID` - no need to update the containers, because they are equal :smile:

Alternatively;

- get all the volumes that are known by the daemon
- for each volume, generate an `ID`
- migrate existing containers and update the volume _name_ reference to volume _id_

> `possibly marking it as "legacy" in the process if we want to preserve that information`. Can you elaborate on this ?

Mainly having "downgrade" in mind; having, e.g. a boolean indicating that a volume was migrated from the "old" storage to "new" storage model, would allow having some migration to revert that change when downgrading.

This would still be a challenge of course, because an older daemon won't be able to deal with the new storage model; possibly we'd have to store both the old and the new format (but I haven't given that much thinking, and I haven't fully dived into the internals of the volume store 😅 )

For the container; running a container that uses a `hey-hello` volume,
I see the volume name is in these locations;

```
"HostConfig": {
    "Binds": [
        "hey-hello:/foo"
    ],
```

and

```
"Mounts": [
    {
        "Type": "volume",
        "Name": "hey-hello",
        "Source": "/var/lib/docker/volumes/hey-hello/_data",
        "Destination": "/foo",
        "Driver": "local",
        "Mode": "z",
        "RW": true,
        "Propagation": ""
    }
],
```

I guess the `Mounts` may get an additional `ID` field (again, want to check), not sure what `Binds` would get. Need to dive into this a bit more :smile:

**Comment by cpuguy83** [comment]
Keep in mind that even renaming containers, which has all state stored completely in the engine, is a huge pain and has historically the cause of many bugs (probably some still exist).

Volumes are a different beast.

Name is for all intents and purposes an ID in Docker. It is entirely unique and cannot be shared amongst multiple drivers. This concept cannot change.

A driver must be aware of name changes.

**Comment by rhoerbe** [comment]
A useful makeshift would be a high-level work-around, which would emulate `docker volume mv` with a clone/remove.

**Comment by ashishxooa** [comment]
+1 to this feature need.

**Comment by den4uk** [comment]
`+ 1` Changing my deployment from swarm to docker-compose, and the volume naming changes. Renaming volumes would solve these issues. A required feature.

**Comment by franklinyu** [comment]
@liuchong Nope; from the dialog above, it’s probably “if you want, do it your own”.

**Comment by Dirc** [comment]
COPY a volume would already be great.

RENAME a volume would be a very nice feature, but seems a little complex/hard especialy the . I think most use cases want just the same volume with a different name, which they can use in new containers. 
Use cases like: switching between single container, Compose, Swarm.

Therefore a much more simple feature like COPY a volume would already be great. Removing an old volume will be a manual task then, which is fine by me.

**Comment by cpfeiffer** [comment]
@Dirc Please don't hijack this issue. Copying a TB volume just to delete it afterwards would certainly *not be great*.

**Comment by liuchong** [comment]
+1

**Comment by Dirc** [comment]
@cpfeiffer You are right, for that use-case you really need the RENAME feature.

**Comment by eaglet3d** [comment]
Just another useful feature requested by the community and ignored by Docker for over a year.  <sigh>

**Comment by cpuguy83** [comment]
@eaglet3d Nobody here is at your beck and call. If someone is interested in this feature, they should submit a proposal to accomplish it. There is not unlimited resources to work on everything.

**Comment by eaglet3d** [comment]
I am not asking you to be at my beck and call.  I do, however, expect that
your team recognize useful features that are proposed by the community and
offer suggestions for how to get it implemented so that everyone can
benefit.

What is your feature request process then?  Do I need to do a pull
request?  If you can't do the work, show me a link to the contributor
documentation on policies and procedures and I will be glad to make a
proposed feature enhancement.

Kind regards,
Ralph

Ralph A. Navarro Jr.
*Navarro Computing LLC <https://navarrocomputing.com>*
Mobile: *+1(508)**287-0190*
AIM,YIM,Skype: eaglet3d

**Comment by cpuguy83** [comment]
@eaglet3d There are multiple responses from maintainers on this thread. Please make sure to read them as this is unfortunately not as simple as changing a name.

Thanks for your willingness to contribute what you'd like to see into this project! The contributing guidelines are here: https://github.com/moby/moby/blob/master/CONTRIBUTING.md

**Comment by hlubovac** [comment]
> @Dirc Please don't hijack this issue. Copying a TB volume just to delete it afterwards would certainly _not be great_.

Copy/clone (assuming specifying target name is included) would actually would be more useful then implementing rename-op (and not implemeting clone-op):
* It wouldn't affect your current volume-management logic;
* It would be useful to those wanting to rename volumes also, as well as to those wanting to clone a complete container runtime.

Whatever the "TB" was...

While at it, probably everything should be cloneable (e.g. containers) - recreating isn't quite the same always.

:)

**Comment by JoyceBabu** [comment]
> Whatever the "TB" was...

I believe he meant Tera Byte.

If you want to rename a 300 GB volume on a 500 GB SSD, `copy` / `clone` will not help.

**Comment by rhoerbe** [comment]
let us have the low-hanging fruit for development and small deployments. It would make just refactoring of volume names much easier.

**Comment by cpuguy83** [comment]
It's not low-hanging.

**Comment by rhoerbe** [comment]
IUC lower-hanging than true rename, with its dependencies to storage drivers.

**Comment by hlubovac** [comment]
RE: the "TB" reference - my bad for not recognizing it; sure, it's a valid point. However, we deal with it with already, with this and with various other systems, by being hit with "operation faile; no space available; choose another destination" error messages. So, it's not a valid excuse, IMO.

**Comment by hlubovac** [comment]
That said, I'm sorry I sound ungrateful and/or negative; I just want this to happen, at least eventually, and not being completely blown off. We copy and rename data files all the time, tiny or large. I don't see it impossible. Sure, maybe it hasn't been thought of in advance and now implementing it is more complicated, but that doesn't make it impossible. :)

**Comment by mcans** [comment]
This would be pretty nice. Not just nice to have. I assume this feature would be a nice way around over the issue with unix socket character limitations on mounted volumes as well.

**Comment by cghislai** [comment]
The easy, high level, workaround of creating a new volume and copying the data can easily be implemented in a docker image. In fact, I remember seeing one docker volume tools image in the <1.12 times, but I could not get my hand on it right now.. It is only 2 or 3 1-line commands that do not need to be run quite often, I don't see any reason to include it in the docker client.

Still, making volumes mutable would be an improvement as it would allow to update driver options/relabel them as well.

**Comment by cghislai** [comment]
Thinking over @thaJeztah and @cpuguy83 comments:
```
Change the volume store, so that each volume has both an ID and a Name
```
```
Name is for all intents and purposes an ID in Docker. It is entirely unique and cannot be shared amongst multiple drivers. This concept cannot change.

A driver must be aware of name changes.
```

I think an elegant solution would be to introduce Aliases for a volume. We can keep the current Name as the unique identifier with backward compatibility.
- Renaming a volume would be adding an alias, as long as it is unique across all existing volumes Name and Aliases
- Removing an alias would not be permitted as it might be referenced elsewhere.
- Removing a volume would free its Name and Aliases.
- Volume could be referenced by their Name or Aliases.
- The client would resolve a volume alias, the API would still only use Name.

Am I missing something?

**Comment by cpuguy83** [comment]
The problem is if we aren’t persisting the alias in the plugin there is little point.
Also aliases will need to be unique as well.

Ultimately I guess it’s fine to add an alias IF it’s persisted to the plugin.
To do so we’ll need to add a new capability check for alias support.

However, I’m not sure this really solves the problem where people want to rename a volume and re-use that name again.
Volumes referenced by some generated ID are not generally helpful to users and are impossible to understand when looking
at the storage backend.
While I did not come up with this idea to use names as they are, it is likely because people need a human reference-able object that
follows through the whole stack.

**Comment by x-yuri** [comment]
You can't rename a volume, but you can copy data. Checkout the following [Server Fault answer][1] (workaround).

[1]: https://serverfault.com/a/989440/162443

**Comment by inbay** [comment]
It would be really valuable to have an alias; $LABEL for each volume which would be used as a symbol to reference the volume identified by its hex value.

This would allow users to have a modifiable, and potentially arbitrary reference point that they could customize to their needs regardless of how the docker itself was initiated, without also complicating the host OS.

While this could be seen as a work-around, it provides value to anyone auditing a docker host, as notes could be added for any number of reasons; antiquation, compatibility, interoperability, requirements, etc.
And because Docker is (always) under development, having some user entered information which can be maintained through an upgrade would be quite valuable.

**Comment by inbay** [comment]
It could be as easy as any script operating with a referenced volume;
receive docker volume identifier

[Finding referenced docker volume]
1] Check against hex values for match
2] If match found use identified volume
3] If no match, check [NEW DOCKER SYMBOL] table
4] If match, use hex value associated to this symbol
5] Else, tell the user no volume could be found with that symbol.

To implement, the metadata attached to volumes could be modified, or a table associated to them such that this value could be added to metadata or as a reference variable.

In the former case it would make the commands a bit more redundant, as it would be necessary to review two tables, but the latter case would be only one, noting of course that this would make that table inherently more complex so additional cost would be imposed even with this case.

**Comment by delanym** [comment]
*bump*

**Comment by FranklinYu** [comment]
I think Docker team made the point clear, but this thread is too long to read. It would be helpful to label this issue with [`good first issue`](https://help.github.com/en/github/building-a-strong-community/encouraging-helpful-contributions-to-your-project-with-labels) or something similar.

**Comment by adamkdean** [comment]
This is still something that, all these years later, would still be really useful. Sad to see it still doesn't exist.

**Comment by PhilippeFerreiraDeSousa** [comment]
Up

**Comment by dk6688** [comment]
worthy feature. my thumbs up.

**Comment by shakudu** [comment]
Surprised to learn this is not a thing yet...

**Comment by thaJeztah** [comment]
Please don't leave "me too" or "+1" comments on this thread to prevent important information captured in the discussion above from being hidden. I'll cleanup some comments above to keep the discussion readable; if you want to stay informed on changes, use the "subscribe" button on the right and add your 👍 in the first comment.

This is (unfortunately) not a trivial change to make (details in comments from myself and @cpuguy83 above outline some reasons). While (can't talk for all) maintainers would be open to someone contributing changes for this enhancements, it _would_ require a good technical design and (migration) plan to prevent breaking changes (taking volume-driver plugins into account).

If a contributor wants to work on this, be sure to read both the thread above, and the discussion on https://github.com/moby/moby/pull/33966#issuecomment-313821356

**Comment by kauedg** [comment]
Suggested behaviour, by means of creating a new field "Id":

```
$ docker volume create 
501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b

$ docker volume inspect 501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b 
[
    {
        "CreatedAt": "2020-06-12T17:54:27-03:00",
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b/_data",
        "Name": "501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b",
        "Id":  "501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b",
        "Options": {},
        "Scope": "local"
    }
]

$ docker volume rename  501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b new_name
new_name

$ docker volume inspect new_name 
[
    {
        "CreatedAt": "2020-06-12T17:54:27-03:00",
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b/_data",
        "Name": "new_name",
        "Id":  "501bad238d8c3a96694a290f0d748863d4a4e4821222df3b719d8f59d7e0ba4b",
        "Options": {},
        "Scope": "local"
    }
]
```

**Comment by cdalexndr** [comment]
My use case:
I want to change the docker-compose folder name prefix (https://github.com/docker/compose/issues/2982).
After prefix change, new volumes are created instead of using old ones (diff prefixed name).
Containers I can re-create (stateless), but persistent data I cannot.
For this rename to work, I require to rename the volumes.... this feature that is missing.

**Comment by mshavliuk** [comment]
This can be a workaround for the time being:

```bash
SOURCE=app_dir
DEST=newapp_dir

docker run --rm \
    -v ${SOURCE}:/var/volume/source:ro \
    -v ${DEST}:/var/volume/dest:rw \
    alpine sh -c "cp -rp /var/volume/source/* /var/volume/dest/"
```

Or if you like to risk you can use `mv` instead of `cp` and remove the original volume after the migration is done:

```bash
SOURCE=app_dir
DEST=newapp_dir

docker run --rm \
    -v ${SOURCE}:/var/volume/source:rw \
    -v ${DEST}:/var/volume/dest:rw \
    alpine sh -c "mv /var/volume/source/* /var/volume/dest/"

docker volume rm ${SOURCE}
```

**Comment by JoyceBabu** [comment]
> My use case:
> 
> 
> 
> I want to change the docker-compose folder name prefix (https://github.com/docker/compose/issues/2982).
> 
> After prefix change, new volumes are created instead of using old ones (diff prefixed name).
> 
> Containers I can re-create (stateless), but persistent data I cannot.
> 
> For this rename to work, I require to rename the volumes.... this feature that is missing.
> 
> 
> 
> Note that I'm only using local volumes.
> 
> 

You can mark the volume as external, as a temporary workaround. But it won't work if you are using scaling.

**Comment by divmgl** [comment]
Is there at least a way to clone the volume? That would somewhat solve this.

Anyhow, [this comment](https://github.com/moby/moby/issues/31154#issuecomment-360531460) worked perfectly.

**Comment by hugovidafe** [comment]
Almost joining march 2023 :D

**Comment by thaJeztah** [comment]
@hugovidafe feel free to contribute; https://github.com/moby/moby/issues/31154#issuecomment-591442092
